### PR TITLE
vscode-extensions.vscjava.vscode-gradle: init at 3.12.2022110200

### PR DESCRIPTION
**Pull request by rhoriguchi**
###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 1]
```diff
@@ -2703,6 +2703,18 @@ let
 
       ms-vsliveshare.vsliveshare = callPackage ./ms-vsliveshare-vsliveshare { };
 
+      vscjava.vscode-gradle = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "vscjava";
+          publisher = "vscode-gradle";
+          version = "3.12.2022110200";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:2710]:
```suggestion
          version = "3.12.2023030400";
```

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 2]
```diff
@@ -2703,6 +2703,18 @@ let
...
+          sha256 = "sha256-9506wTeXw+MfsSQl9XZJ5gmPM9+wx39y0eGRhAkMhT8=";
+        };
+        meta = {
+          license = lib.licenses.mit;
+        };
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:2715]: Please also fill meta with missing parameters: downloadPage, homepage, changelog, description, etc.

**Comment by superherointj** [comment]
Please fix merge conflict.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 3]
```diff
@@ -2953,6 +2953,22 @@ let
         };
       };
 
+      vscjava.vscode-gradle = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "vscjava";
+          publisher = "vscode-gradle";
+          version = "3.12.2023032100";
+          sha256 = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:2961]:
```suggestion
          sha256 = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
```
This is wrong/empty.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 4]
```diff
@@ -2953,6 +2953,22 @@ let
...
+        };
+        meta = {
+          changelog = "https://marketplace.visualstudio.com/items/vscjava.vscode-gradle/changelog";
+          description = "Visual Studio Code Extension - visual interface for your Gradle build";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:2965]:
```suggestion
          description = "A Visual Studio Code extension for Gradle build tool";
```

**Comment by superherointj** [comment]
Consider adding yourself as maintainer to receive notifications on issues pertaining this extension that you use.

File: pkgs/applications/editors/vscode/extensions/default.nix (hunk 1 above)
Comment (rhoriguchi) [reply pkgs/applications/editors/vscode/extensions/default.nix]: I'm struggling to download this extension with `buildVscodeMarketplaceExtension`. If I get it from VS Code marketplace Version History tab I can download it with no issues from <https://marketplace.visualstudio.com/_apis/public/gallery/publishers/vscjava/vsextensions/vscode-gradle/3.12.2023032100/vspackage>.

```console
> nix repl '<nixpkgs>'
> :b pkgs.vscode-utils.buildVscodeMarketplaceExtension {
  mktplcRef = {
    name = "vscjava";
    publisher = "vscode-gradle";
    version = "3.12.2023032100";
    sha256 = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
  };
}

error: builder for '/nix/store/1d2paxk03rziz1bi26yida6fiwm49n3r-vscode-gradle-vscjava.zip.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://vscode-gradle.gallery.vsassets.io/_apis/public/gallery/publisher/vscode-gradle/extension/vscjava/3.12.2023032100/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0   319    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download vscode-gradle-vscjava.zip from any mirror
       For full logs, run 'nix log /nix/store/1d2paxk03rziz1bi26yida6fiwm49n3r-vscode-gradle-vscjava.zip.drv'.
error: 1 dependencies of derivation '/nix/store/c2v8cy7xcldr7q2nwajbgy8c28mfm5qa-vscode-extension-vscode-gradle-vscjava-3.12.2023032100.drv' failed to build
```

File: pkgs/applications/editors/vscode/extensions/default.nix (hunk 3 above)
Comment (rhoriguchi) [reply pkgs/applications/editors/vscode/extensions/default.nix:2961]: Please read https://github.com/NixOS/nixpkgs/pull/199157#discussion_r1145507322, was hoping i can get the hash from ofborg.

File: pkgs/applications/editors/vscode/extensions/default.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/applications/editors/vscode/extensions/default.nix]: @ShamrockLee Any idea?

Comment (ShamrockLee) [reply pkgs/applications/editors/vscode/extensions/default.nix]: The VSCode extension gallery API is left undocumented. The only publicly-available "documentation" is the source code of VSCode OSS.

I guess they have changed the URL used to download the extension file (`.vsix`), and could add the new URL into the `urls` list of the fetcher.

However, I constantly got blocked:

```json
{
  "$id": "1",
  "innerException": null,
  "message": "Request was blocked due to exceeding usage of resource Count in namespace AnonymousId. For more information on why your request was blocked, see the topic \"Rate limits\" on the Microsoft Web site (https://go.microsoft.com/fwlink/?LinkId=823950).",
  "typeName": "Microsoft.TeamFoundation.Framework.Server.RequestBlockedException, Microsoft.TeamFoundation.Framework.Server",
  "typeKey": "RequestBlockedException",
  "errorCode": 0,
  "eventId": 3000
}
```

and all my attempts to use `curl` and `wget` to download from this url failed. (My knowledge toward cURL and crowlers are still limited. Maybe there's a way to satisfy the site?)

Comment (ShamrockLee) [reply pkgs/applications/editors/vscode/extensions/default.nix]: @AmeerTaweel Have you encountered similar issues (packages like `vscode-gradle.vscjava` failed to fetch) on your [nix-vscode-extensions](https://github.com/nix-community/nix-vscode-extensions) project?

Comment (superherointj) [reply pkgs/applications/editors/vscode/extensions/default.nix]: @ShamrockLee Can't we just package extensions from sources? Looks ideal. Thoughts?

Comment (superherointj) [reply pkgs/applications/editors/vscode/extensions/default.nix]:
```
vscjava.vscode-gradle = buildVscodeMarketplaceExtension rec {
        mktplcRef = {
          publisher = "vscjava";
          name = "vscode-gradle";
          version = "3.12.6";
        };
        vsix = fetchurl {
          name = "vscjava.vscode-gradle.zip";
          url = "https://github.com/microsoft/vscode-gradle/releases/download/${mktplcRef.version}/${mktplcRef.publisher}.${mktplcRef.name}-${mktplcRef.version}.vsix";
          hash = "sha256-j4JyhNGsRlJmS8Wj38gLpC1gXVvdPx10cgzP8dXmmNo=";
        };
};
```

@rhoriguchi See if this works for you.
@ShamrockLee Any improvement on format?

Comment (superherointj) [reply pkgs/applications/editors/vscode/extensions/default.nix]: Whenever available, we could make use of VSIX asset by default. What do you think?

Comment (ShamrockLee) [reply pkgs/applications/editors/vscode/extensions/default.nix]: @superherointj Looks good by eyeball (not teted yet).

Comment (rhoriguchi) [reply pkgs/applications/editors/vscode/extensions/default.nix]: Sorry @superherointj and @ShamrockLee I'm an idiot and can't read, for some reason I switched `name` and `publisher` while making my changes 🤦🏻‍♂️.

**Review by superherointj — APPROVED** [review]

**Pull request merged**